api/receiver/appengine.go

```diff
@@ -57,12 +58,24 @@ func NewAppEngineAPI(ctx context.Context) AppEngineAPI {
 	return &appEngineAPIImpl{
 		AppEngineAPIImpl: shared.NewAppEngineAPI(ctx),
 		queue:            ResultsQueue,
+		store:            shared.NewAppEngineDatastore(ctx, false),
 	}
 }
 
 func (a *appEngineAPIImpl) addTestRun(testRun *shared.TestRun) (*DatastoreKey, error) {
-	key := datastore.NewIncompleteKey(a.Context(), "TestRun", nil)
-	key, err := datastore.Put(a.Context(), key, testRun)
+	ctx := a.Context()
+	var key *datastore.Key
+	if testRun.ID > 0 {
+		key = datastore.NewKey(ctx, "TestRun", "", testRun.ID, nil)
+		err := datastore.Get(ctx, key, new(shared.TestRun))
```

Comment:
This needs to go into a transaction: https://cloud.google.com/datastore/docs/concepts/entities#creating_an_entity (see `insert` an entity)

While doing so, let's make `Insert` a new method on the `shared.Datastore` interface.

Reply:
I disagree that this needs to be in a transaction; we're not doing multiple actions. Either the entity doesn't exist (so, nothing happens), or it does and we return. We don't ever do both a real get/put.

Reply:
There's a `Put` right after the whole `if` block. The race condition would be:

Two parallel requests trying to create two test runs with the same ID (bugs in the code or retries, etc.). Two requests may both think the ID doesn't exist, and they'll simultaneously try to write to that key. One test run will be lost that way, and in case the runs are different, we have data loss.

Reply:
Oh, I see. Your contrived error relies on some incorrect code somewhere. Done.